### PR TITLE
[neox-2.x] add mpt

### PR DESCRIPTION
**Pull request by KickSeason**
### Step 2/6  from `neo-ngd/neo` to `neox-2.x`
Basically same as #1442 
* keep `IKVStore` and `IKVReadOnlyStore` because `ISnapshot` cant be easily used.
* keep `Skip` `Concat` because linq is slow

@erikzhang @shargon Please have a look.

### CODE REVIEW

File: neo/IO/ByteArrayEqualityComparer.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+            for (int i = 0; i < a.Length; i++)
+            {
+                if (a[i] != b[i]) return false;
+            }
+            return true;
+        }
+
+        public int GetHashCode(byte[] a)
+        {
+            return a.ToInt32(0);
```
Comment (shargon) [inline neo/IO/ByteArrayEqualityComparer.cs:22]: It will produce an error if the length is less than 4

File: neo/Trie/MPT/MPTNode.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        public MPTNode()
+        {
+            Dirty = true;
+        }
+
+        public byte[] Encode()
+        {
+            MemoryStream ms = null;
+            try
+            {
```
Comment (shargon) [inline neo/Trie/MPT/MPTNode.cs:42]: Copy the changes from neo3 branch

File: neo/Trie/MPT/MPTNode/BranchNode.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+using Neo.IO.Json;
+using System.IO;
+
+namespace Neo.Trie.MPT
+{
+    public class BranchNode : MPTNode
+    {
+        public const int ChildCount = 17;
+        public MPTNode[] Children = new MPTNode[ChildCount];
+
```
Comment (shargon) [inline neo/Trie/MPT/MPTNode/BranchNode.cs:11]: Remove empty line

File: neo/Trie/MPT/MPTNode/ExtensionNode.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+using Neo.Cryptography;
+using Neo.IO;
+using Neo.IO.Json;
+using System.IO;
+
+namespace Neo.Trie.MPT
+{
+    public class ExtensionNode : MPTNode
+    {
+        //max StorageKey lenght
```
Comment (shargon) [inline neo/Trie/MPT/MPTNode/ExtensionNode.cs:10]:
```suggestion
        // Max StorageKey length
```

File: neo/Trie/MPT/MPTNode/ExtensionNode.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        public override void EncodeSpecific(BinaryWriter writer)
+        {
+            writer.WriteVarBytes(Key);
+            var hashNode = new HashNode(Next.GetHash());
+            hashNode.EncodeSpecific(writer);
+        }
+
+        public override void DecodeSpecific(BinaryReader reader)
+        {
+            Key = reader.ReadVarBytes();
```
Comment (shargon) [inline neo/Trie/MPT/MPTNode/ExtensionNode.cs:29]:
```suggestion
            Key = reader.ReadVarBytes(MaxKeyLength);
```

File: neo/Trie/MPT/MPTNode/LeafNode.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        public override void EncodeSpecific(BinaryWriter writer)
+        {
+            writer.WriteVarBytes(Value);
+        }
+
+        public override void DecodeSpecific(BinaryReader reader)
+        {
+            Value = reader.ReadVarBytes();
```
Comment (shargon) [inline neo/Trie/MPT/MPTNode/LeafNode.cs:30]:
```suggestion
            Value = reader.ReadVarBytes(ushort.MaxValue);
```

File: neo/Trie/MPT/MPTNodeType.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+namespace Neo.Trie.MPT
+{
+    public enum NodeType
+    {
+        BranchNode = 0x00,
+        ExtensionNode = 0x01,
+        HashNode = 0x02,
+        LeafNode = 0x03,
+    }
+}
```
Comment (shargon) [inline neo/Trie/MPT/MPTNodeType.cs:10]: Blank line at the end

File: neo/Trie/MPT/MPTProofDb.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+using Neo.IO;
+using Neo.Cryptography;
+using System.Collections.Generic;
+
+namespace Neo.Trie.MPT
+{
+    public class MPTProofStore : IKVReadOnlyStore
+    {
+        private Dictionary<byte[], byte[]> store = new Dictionary<byte[], byte[]>(ByteArrayEqualityComparer.Default);
+        public MPTProofStore(HashSet<byte[]> proof)
```
Comment (shargon) [inline neo/Trie/MPT/MPTProofDb.cs:10]: Please, add one enter between variables and methods.

File: neo/Trie/MPT/MPTReadOnlyTrie.cs [hunk 9]
```diff
@@ -0,0 +1,144 @@
+using Neo.IO;
+using System;
+using System.Collections.Generic;
+
+namespace Neo.Trie.MPT
+{
+    public class MPTReadOnlyTrie
+    {
+        private MPTReadOnlyDb rodb;
```
Comment (shargon) [inline neo/Trie/MPT/MPTReadOnlyTrie.cs:9]:
```suggestion
        private readonly MPTReadOnlyDb rodb;
```

File: neo/Trie/MPT/MPTTrie.cs [hunk 10]
```diff
@@ -0,0 +1,250 @@
+using Neo.IO.Json;
+using Neo;
+using System;
+
+namespace Neo.Trie.MPT
+{
+    public class MPTTrie : MPTReadOnlyTrie
+    {
+        public static uint Version = 0;
+        private MPTDb db;
```
Comment (shargon) [inline neo/Trie/MPT/MPTTrie.cs:10]:
```suggestion
        private readonly MPTDb db;
```

**Comment by KickSeason** [comment]
@erikzhang @shargon I still have 2 prs that depend on this, hope you can review this pr first. 🙏

File: neo/Trie/MPT/MPTNode.cs [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        public MPTNode()
+        {
+            Dirty = true;
+        }
+
+        public byte[] Encode()
+        {
+            using (BinaryWriter writer = new BinaryWriter(new MemoryStream(), Encoding.UTF8, true))
```
Comment (shargon) [inline neo/Trie/MPT/MPTNode.cs:40]: LeaveOpen must be false
```suggestion
            using (BinaryWriter writer = new BinaryWriter(new MemoryStream(), Encoding.UTF8, false))
```

File: neo/Trie/MPT/MPTProofDb.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
+using Neo.IO;
+using Neo.Cryptography;
+using System.Collections.Generic;
+
+namespace Neo.Trie.MPT
+{
+    public class MPTProofStore : IKVReadOnlyStore
+    {
+        private Dictionary<byte[], byte[]> store = new Dictionary<byte[], byte[]>(ByteArrayEqualityComparer.Default);
```
Comment (shargon) [inline neo/Trie/MPT/MPTProofDb.cs:9]: This doesn't need to be persistent?

File: neo/Trie/MPT/MPTReadOnlyTrie.cs [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+                        node = newNode;
+                        return TryGet(ref node, path, out value);
+                    }
+                case BranchNode branchNode:
+                    {
+                        if (path.Length == 0)
+                        {
+                            return TryGet(ref branchNode.Children[16], path, out value);
+                        }
+                        return TryGet(ref branchNode.Children[path[0]], path.Skip(1), out value);
```
Comment (shargon) [inline neo/Trie/MPT/MPTReadOnlyTrie.cs:66]: Could you add comments about why this skip and offsets?

File: neo/Trie/MPT/MPTProofDb.cs (hunk 12 above)
Comment (KickSeason) [reply neo/Trie/MPT/MPTProofDb.cs:9]: `ProofStore` is only used when verify the proofs.  Proofs, root and key are provided by users. `VerifyProof` only uses the trie logic, like rpc `invokeScript`.
I can remove it if you don't like.

File: neo/Trie/MPT/MPTReadOnlyTrie.cs (hunk 13 above)
Comment (KickSeason) [reply neo/Trie/MPT/MPTReadOnlyTrie.cs:66]: canvas
Take this graph for example.
If the path is `0a0c` you get `val` and the value is `v1`
If the path is longer than `0a0c` like `0a0c0?*`,  `?` is from 0 to f, we find the value from the corresponding branch.

**Comment by KickSeason** [comment]
@shargon Can we merge this pr?

**Comment by shargon** [comment]
It could be merged, the code looks good, but should be tested more, but it's in a different branch so there is no problem with this.

**Pull request merged**